### ISSUE TITLE
More fuzz tests

### ISSUE THREAD

**Issue by uncomputable** [BlockstreamResearch/rust-simplicity#116]
We keep finding bugs via fuzz tests, mostly thanks to @apoelstra running local tests. We should do this more on a crate level. Maybe (almost) every function has one or many fuzz tests associated with it?

We should also explore new ways how we use fuzz tests. Currently we mostly look for crashes. Alternatively, functions could have pre and post conditions, similar to the C code, that we fuzz against. The fuzzer produces an input, we filter for inputs that satisfy the pre condition, and we check if the post condition is fulfilled.

**Comment by apoelstra** [comment]
Agreed, but right now we are blocked on `bitcoin_hashes` breaking when compiled with `cfg(fuzzing)`. Fixing this properly will require a sequence of new releases, but it might make sense for us to just fork bitcoin_hashes 0.11 for now, remove the fuzzing behavior, then add a `[patch]` entry to our `fuzz/Cargo.toml`.

**Comment by uncomputable** [comment]
> Value encoding roundtrips are candidates for future fuzz tests

see #266
